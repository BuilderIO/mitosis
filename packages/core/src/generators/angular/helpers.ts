import { type MitosisComponent } from '@/types/mitosis-component';

export const HELPER_FUNCTIONS = (
  isTs?: boolean,
): {
  [key: string]: string;
} => ({
  useObjectWrapper: `useObjectWrapper(...args${isTs ? ': any[]' : ''}) {
    let obj = {}
    args.forEach((arg) => {
      obj = { ...obj, ...arg };
    });
    return obj;
  }`,
  useObjectDotValues: `useObjectDotValues(obj${isTs ? ': any' : ''})${isTs ? ': any[]' : ''}) {
    return Object.values(obj);
  }`,
  useTypeOf: `useTypeOf(obj${isTs ? ': any' : ''})${isTs ? ': string' : ''}) {
    return typeof obj;
  }`,
  useJsonStringify: `useJsonStringify(...args${isTs ? ': any' : ''})${isTs ? ': string' : ''}) {
    return JSON.stringify(...args);
  }`,
  setAttributes: `setAttributes(el${isTs ? ': HTMLElement' : ''}, value${
    isTs ? ': any' : ''
  }, changes${isTs ? '?: any' : ''}) {
    if (!el) {
      return;
    }
    const target = typeof changes === 'undefined' ? value : changes;
    Object.keys(target).forEach((key) => {
      if (key.startsWith('on')) {
        if (this._listenerFns.has(key)) {
          this._listenerFns.get(key)${isTs ? '!' : ''}();
        }
        this._listenerFns.set(key, this.renderer.listen(
          el,
          key.replace('on', '').toLowerCase(),
          target[key]
        ));
      } else {
        this.renderer.setAttribute(el, key.toLowerCase(), target[key] ?? '');
      }
    });
  }`,
});

export const getAppropriateTemplateFunctionKeys = (code: string) =>
  Object.keys(HELPER_FUNCTIONS()).filter((key) => code.includes(key));

/**
 * Adds code to the `onUpdate` hook of a MitosisComponent.
 *
 * @param {MitosisComponent} root - The root MitosisComponent.
 * @param {string} code - The code to be added to the `onUpdate` hook.
 */
export const addCodeToOnUpdate = (root: MitosisComponent, code: string) => {
  root.hooks.onUpdate = root.hooks.onUpdate || [];
  root.hooks.onUpdate.push({
    code,
  });
};

/**
 * Adds code to the `onInit` hook of a MitosisComponent.
 *
 * @param {MitosisComponent} root - The root MitosisComponent.
 * @param {string} code - The code to be added to the `onInit` hook.
 */
export const addCodeToOnInit = (root: MitosisComponent, code: string) => {
  if (!root.hooks.onInit?.code) {
    root.hooks.onInit = { code: '' };
  }
  root.hooks.onInit.code += `\n${code};`;
};

/**
 * Creates a reactive state in Angular.
 * Initializes the state with `null` because we cannot access `state.` or `props.` properties before the component is initialized.
 * Adds the code (init/re-init code) to the `onInit` and `onUpdate` hooks.
 * @param root The root MitosisComponent.
 * @param stateName The name of the reactive state.
 * @param code The code to be added to the onInit and onUpdate hooks.
 */
export const makeReactiveState = (root: MitosisComponent, stateName: string, code: string) => {
  root.state[stateName] = { code: 'null', type: 'property' };
  addCodeToOnInit(root, code);
  addCodeToOnUpdate(root, code);
};
