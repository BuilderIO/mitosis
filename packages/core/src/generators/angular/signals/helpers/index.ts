export const getAngularCoreImportsAsString = ({
  refs,
  output,
  input,
  model,
  onPush,
  effect,
  signal,
}: {
  refs: boolean;
  output: boolean;
  input: boolean;
  model: boolean;
  onPush: boolean;
  effect: boolean;
  signal: boolean;
}): string => {
  const angularCoreImports: Record<string, boolean> = {
    Component: true,
    AfterViewInit: true,
    viewChild: refs,
    ElementRef: refs,
    model,
    output,
    input,
    effect,
    signal,
    ChangeDetectionStrategy: onPush,
  };
  return Object.entries(angularCoreImports)
    .map(([key, bool]) => (bool ? key : ''))
    .filter((key) => !!key)
    .join(', ');
};
