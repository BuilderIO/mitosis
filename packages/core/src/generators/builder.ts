import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';
import { BuilderElement } from '@builder.io/sdk';
import { getStateObjectString } from '../helpers/get-state-object-string';
import { fastClone } from '../helpers/fast-clone';
import dedent from 'dedent';
import { format } from 'prettier/standalone';
import json5 from 'json5';
import { isUpperCase } from '../helpers/is-upper-case';
import { mediaQueryRegex, sizes } from '../constants/media-sizes';
import { filterEmptyTextNodes } from '../helpers/filter-empty-text-nodes';
import { isComponent } from '../helpers/is-component';

const builderBlockPrefixes = ['Amp', 'Core', 'Builder', 'Raw', 'Form'];
const mapComponentName = (name: string) => {
  if (name === 'CustomCode') {
    return 'Custom Code';
  }
  for (const prefix of builderBlockPrefixes) {
    if (name.startsWith(prefix)) {
      const suffix = name.replace(prefix, '');
      if (isUpperCase(suffix[0])) {
        return `${prefix}:${name.replace(prefix, '')}`;
      }
    }
  }
  return name;
};

const componentMappers: {
  [key: string]: (
    node: JSXLiteNode,
    options: ToBuilderOptions,
  ) => BuilderElement;
} = {
  Columns(node, options) {
    const block = blockToBuilder(node, options, { skipMapper: true });

    const columns = block.children!.map((item) => ({
      blocks: item.children,
    }));

    block.component!.options.columns = columns;

    block.children = [];

    return block;
  },
  For(node, options) {
    return el(
      {
        component: {
          name: 'Fragment',
        },
        repeat: {
          collection: node.bindings.each as string,
          itemName: node.bindings._forName as string,
        },
        children: node.children
          .filter(filterEmptyTextNodes)
          .map((node) => blockToBuilder(node, options)),
      },
      options,
    );
  },
  Show(node, options) {
    return el(
      {
        // TODO: the reverse mapping for this
        component: {
          name: 'Fragment',
        },
        bindings: {
          show: node.bindings.when as string,
        },
        children: node.children
          .filter(filterEmptyTextNodes)
          .map((node) => blockToBuilder(node, options)),
      },
      options,
    );
  },
};

const el = (
  options: Partial<BuilderElement>,
  toBuilderOptions: ToBuilderOptions,
): BuilderElement => ({
  '@type': '@builder.io/sdk:Element',
  ...(toBuilderOptions.includeIds && {
    id:
      'builder-' +
      Math.random()
        .toString(36)
        .split('.')[1],
  }),
  ...options,
});

export type ToBuilderOptions = {
  includeIds?: boolean;
};

function tryFormat(code: string) {
  let str = code;
  try {
    str = format(str, {
      parser: 'babel',
      plugins: [
        require('prettier/parser-babel'), // To support running in browsers
      ],
    });
  } catch (err) {
    console.error('Format error for code:', str);
    throw err;
  }
  return str;
}

type InternalOptions = {
  skipMapper?: boolean;
};

export const blockToBuilder = (
  json: JSXLiteNode,
  options: ToBuilderOptions = {},
  _internalOptions: InternalOptions = {},
): BuilderElement => {
  const mapper = !_internalOptions.skipMapper && componentMappers[json.name];
  if (mapper) {
    return mapper(json, options);
  }
  if (json.properties._text || json.bindings._text) {
    return el(
      {
        tagName: 'span',
        // responsiveStyles: {
        //   large: json.properties.css as any,
        // },
        bindings: {
          // TODO: css to responsiveStyles and back
          // ...(json.bindings as any),
          ...(json.bindings._text
            ? {
                'component.options.text': json.bindings._text as string,
                'json.bindings._text': undefined as any,
              }
            : {}),
        },
        component: {
          name: 'Text',
          options: {
            text: json.properties._text,
          },
        },
      },
      options,
    );
  }

  const thisIsComponent = isComponent(json);

  let bindings = thisIsComponent ? {} : json.bindings;
  if (thisIsComponent) {
    for (const key in json.bindings) {
      bindings[`component.options.${key}`] = json.bindings[key];
    }
  }

  const hasCss = !!json.bindings.css;
  let responsiveStyles: {
    large: Partial<CSSStyleDeclaration>;
    medium?: Partial<CSSStyleDeclaration>;
    small?: Partial<CSSStyleDeclaration>;
  } = {
    large: {},
  };
  if (hasCss) {
    const cssRules = json5.parse(json.bindings.css as string);
    const cssRuleKeys = Object.keys(cssRules);
    for (const ruleKey of cssRuleKeys) {
      const mediaQueryMatch = ruleKey.match(mediaQueryRegex);
      if (mediaQueryMatch) {
        const [fullmatch, pixelSize] = mediaQueryMatch;
        const sizeForWidth = sizes.getSizeForWidth(Number(pixelSize));
        const currentSizeStyles = responsiveStyles[sizeForWidth] || {};
        responsiveStyles[sizeForWidth] = {
          ...currentSizeStyles,
          ...cssRules[ruleKey],
        };
      } else {
        responsiveStyles.large = {
          ...responsiveStyles.large,
          [ruleKey]: cssRules[ruleKey],
        };
      }
    }
  }

  return el(
    {
      tagName: thisIsComponent ? undefined : json.name,
      ...(hasCss && {
        responsiveStyles,
      }),
      ...(thisIsComponent && {
        component: {
          name: mapComponentName(json.name),
          options: json.properties,
        },
      }),
      properties: thisIsComponent ? undefined : (json.properties as any),
      bindings: thisIsComponent ? undefined : (json.bindings as any),
      children: json.children
        .filter(filterEmptyTextNodes)
        .map((child) => blockToBuilder(child, options)),
    },
    options,
  );
};

export const componentToBuilder = (
  componentJson: JSXLiteComponent,
  options: ToBuilderOptions = {},
) => {
  return fastClone({
    data: {
      jsCode: tryFormat(dedent`
        Object.assign(state, ${getStateObjectString(componentJson)});
      `),
      blocks: componentJson.children
        .filter(filterEmptyTextNodes)
        .map((child) => blockToBuilder(child, options)),
    },
  });
};
