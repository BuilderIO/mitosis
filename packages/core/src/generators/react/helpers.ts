import { isMitosisNode } from '@/helpers/is-mitosis-node';
import { stripStateAndPropsRefs } from '@/helpers/strip-state-and-props-refs';
import { MitosisComponent } from '@/types/mitosis-component';
import { MitosisNode } from '@/types/mitosis-node';
import { upperFirst } from 'lodash';
import traverse from 'traverse';

import { ToReactOptions } from './types';

export const processBinding = (str: string, options: ToReactOptions) => {
  if (options.stateType !== 'useState') {
    return str;
  }

  return stripStateAndPropsRefs(str, {
    includeState: true,
    includeProps: false,
  });
};

export const openFrag = (options: ToReactOptions) => getFragment('open', options);
export const closeFrag = (options: ToReactOptions) => getFragment('close', options);
export function getFragment(type: 'open' | 'close', options: ToReactOptions) {
  const tagName = options.preact ? 'Fragment' : '';
  return type === 'open' ? `<${tagName}>` : `</${tagName}>`;
}
export const wrapInFragment = (json: MitosisComponent | MitosisNode) => json.children.length !== 1;

function getRefName(path: string) {
  return upperFirst(path) + 'Ref';
}

export function getCode(str: string = '', options: ToReactOptions): string {
  return processBinding(str, options);
}

export function processTagReferences(json: MitosisComponent, options: ToReactOptions) {
  const namesFound = new Set<string>();

  traverse(json).forEach((el) => {
    if (!isMitosisNode(el)) {
      return;
    }

    const processedRefName = processBinding(el.name, options);

    if (el.name.includes('state.')) {
      switch (json.state[processedRefName]?.type) {
        case 'getter':
          const refName = getRefName(processedRefName);
          if (!namesFound.has(el.name)) {
            namesFound.add(el.name);
            json.hooks.init = {
              ...json.hooks.init,
              code: `
            ${json.hooks.init?.code || ''}
            const ${refName} = ${el.name};
            `,
            };
          }

          el.name = refName;
          break;

        // NOTE: technically, it should be impossible for the tag to be a method or a function in Mitosis JSX syntax,
        // as that will fail JSX parsing.
        case 'method':
        case 'function':

        case 'property':
          const capitalizedName = upperFirst(processedRefName);

          if (capitalizedName !== processedRefName) {
            el.name = capitalizedName;
            json.state[capitalizedName] = { ...json.state[processedRefName]! };

            delete json.state[processedRefName];
          } else {
            el.name = processedRefName;
          }

          break;
      }
    } else {
      el.name = processedRefName;
    }
  });
}
