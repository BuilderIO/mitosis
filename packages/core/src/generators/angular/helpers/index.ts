import { blockToAngular } from '@/generators/angular/classic/blocks';
import { AngularBlockOptions, ToAngularOptions } from '@/generators/angular/types';
import { indent } from '@/helpers/indent';
import { isMitosisNode } from '@/helpers/is-mitosis-node';
import { replaceNodes } from '@/helpers/replace-identifiers';
import {
  DO_NOT_USE_VARS_TRANSFORMS,
  stripStateAndPropsRefs,
} from '@/helpers/strip-state-and-props-refs';
import { nodeHasCss } from '@/helpers/styles/helpers';
import { type MitosisComponent } from '@/types/mitosis-component';
import { MitosisNode } from '@/types/mitosis-node';
import * as babel from '@babel/core';
import { pipe } from 'fp-ts/function';
import traverse from 'neotraverse/legacy';
import { blockToAngularSignals } from '../signals/blocks';

export const HELPER_FUNCTIONS = (
  isTs?: boolean,
): {
  [key: string]: string;
} => ({
  useObjectWrapper: `useObjectWrapper(...args${isTs ? ': any[]' : ''}) {
    let obj = {}
    args.forEach((arg) => {
      obj = { ...obj, ...arg };
    });
    return obj;
  }`,
  useObjectDotValues: `useObjectDotValues(obj${isTs ? ': any' : ''})${isTs ? ': any[]' : ''}) {
    return Object.values(obj);
  }`,
  useTypeOf: `useTypeOf(obj${isTs ? ': any' : ''})${isTs ? ': string' : ''}) {
    return typeof obj;
  }`,
  useJsonStringify: `useJsonStringify(...args${isTs ? ': any' : ''})${isTs ? ': string' : ''}) {
    return JSON.stringify(...args);
  }`,
  setAttributes: `setAttributes(el${isTs ? ': HTMLElement' : ''}, value${
    isTs ? ': any' : ''
  }, changes${isTs ? '?: any' : ''}) {
    if (!el) {
      return;
    }
    const target = typeof changes === 'undefined' ? value : changes;
    Object.keys(target).forEach((key) => {
      if (key.startsWith('on')) {
        if (this._listenerFns.has(key)) {
          this._listenerFns.get(key)${isTs ? '!' : ''}();
        }
        this._listenerFns.set(key, this.renderer.listen(
          el,
          key.replace('on', '').toLowerCase(),
          target[key]
        ));
      } else {
        this.renderer.setAttribute(el, key.toLowerCase(), target[key] ?? '');
      }
    });
  }`,
});

export const getAppropriateTemplateFunctionKeys = (code: string) =>
  Object.keys(HELPER_FUNCTIONS()).filter((key) => code.includes(key));

export const getDefaultProps = ({ defaultProps }: MitosisComponent) => {
  if (!defaultProps) return '';
  const defalutPropsString = Object.keys(defaultProps)
    .map((prop) => {
      const value = defaultProps!.hasOwnProperty(prop) ? defaultProps![prop]?.code : 'undefined';
      return `${prop}: ${value}`;
    })
    .join(',');
  return `const defaultProps: any = {${defalutPropsString}};\n`;
};

/**
 * if any state "property" is trying to access state.* or props.*
 * then we need to move them to onInit where they can be accessed
 * @param json The MitosisComponent.
 */
export const transformState = (json: MitosisComponent) => {
  Object.entries(json.state)
    .reverse()
    .forEach(([key, value]) => {
      if (value?.type === 'property') {
        if (value.code && (value.code.includes('state.') || value.code.includes('props.'))) {
          const code = stripStateAndPropsRefs(value.code, { replaceWith: 'this' });
          json.state[key]!.code = 'null';
          if (!json.hooks.onInit?.code) {
            json.hooks.onInit = { code: '' };
          }
          json.hooks.onInit.code = `\nthis.${key} = ${code};\n${json.hooks.onInit.code}`;
        }
      }
    });
};

/**
 * Checks if the first child has a "key" attribute - used for "For" elements
 * @param node The node which should be "For"
 */
export const hasFirstChildKeyAttribute = (node: MitosisNode): boolean => {
  if (!node.children || node.children.length === 0) {
    return false;
  }

  const firstChildBinding = node.children[0].bindings;
  return Boolean(firstChildBinding && firstChildBinding.key?.code);
};

export const preprocessCssAsJson = (json: MitosisComponent) => {
  traverse(json).forEach((item) => {
    if (isMitosisNode(item)) {
      if (nodeHasCss(item)) {
        if (item.bindings.css?.code?.includes('&quot;')) {
          item.bindings.css.code = item.bindings.css.code.replace(/&quot;/g, '"');
        }
      }
    }
  });
};

export const generateNgModule = (
  content: string,
  name: string,
  componentsUsed: string[],
  component: MitosisComponent,
  bootstrapMapper: Function | null | undefined,
): string => {
  return `import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";

${content}

@NgModule({
  declarations: [${name}],
  imports: [CommonModule${
    componentsUsed.length ? ', ' + componentsUsed.map((comp) => `${comp}Module`).join(', ') : ''
  }],
  exports: [${name}],
  ${bootstrapMapper ? bootstrapMapper(name, componentsUsed, component) : ''}
})
export class ${name}Module {}`;
};

export const traverseToGetAllDynamicComponents = (
  json: MitosisComponent,
  options: ToAngularOptions,
  blockOptions: AngularBlockOptions,
  api: 'signals' | 'classic' = 'classic',
) => {
  const components: Set<string> = new Set();
  let dynamicTemplate = '';
  traverse(json).forEach((item) => {
    if (isMitosisNode(item) && item.name.includes('.') && item.name.split('.').length === 2) {
      const children = item.children
        .map((child) =>
          api === 'classic'
            ? blockToAngular({ root: json, json: child, options, blockOptions })
            : blockToAngularSignals({ root: json, json: child, options, blockOptions }),
        )
        .join('\n');
      dynamicTemplate = `<ng-template #${
        item.name.split('.')[1].toLowerCase() + 'Template'
      }>${children}</ng-template>`;
      components.add(item.name);
    }
  });
  return {
    components,
    dynamicTemplate,
  };
};

export const getTemplateFormat = (template: string): string =>
  indent(template, 8).replace(/`/g, '\\`').replace(/\$\{/g, '\\${');

export const traverseAndCheckIfInnerHTMLIsUsed = (json: MitosisComponent) => {
  let innerHTMLIsUsed = false;
  traverse(json).forEach((item) => {
    if (isMitosisNode(item)) {
      Object.keys(item.bindings).forEach((key) => {
        if (key === 'innerHTML') {
          innerHTMLIsUsed = true;
          return;
        }
      });
    }
  });
  return innerHTMLIsUsed;
};

const { types } = babel;

/**
 * Prefixes state identifiers with this.
 * e.g. state.foo --> this.foo
 */
const prefixState = (code: string): string => {
  return replaceNodes({
    code,
    nodeMaps: [
      {
        from: types.identifier('state'),
        to: types.thisExpression(),
      },
    ],
  }).trim();
};

export const processAngularCode =
  ({
    contextVars,
    outputVars,
    domRefs,
    replaceWith,
  }: {
    contextVars: string[];
    outputVars: string[];
    domRefs: string[];
    replaceWith?: string;
  }) =>
  (code: string) =>
    pipe(
      DO_NOT_USE_VARS_TRANSFORMS(code, {
        contextVars,
        domRefs,
        outputVars,
      }),
      /**
       * Only prefix state that is in the Angular class component.
       * Do not prefix state referenced in the template
       */
      replaceWith === 'this' ? prefixState : (x) => x,
      (newCode) => stripStateAndPropsRefs(newCode, { replaceWith }),
    );
