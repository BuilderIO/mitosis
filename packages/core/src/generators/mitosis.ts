import dedent from 'dedent';
import json5 from 'json5';
import { format } from 'prettier/standalone';
import { Transpiler } from '..';
import { fastClone } from '../helpers/fast-clone';
import { getComponents } from '../helpers/get-components';
import { getRefs } from '../helpers/get-refs';
import { getStateObjectStringFromComponent } from '../helpers/get-state-object-string';
import { mapRefs } from '../helpers/map-refs';
import { renderPreComponent } from '../helpers/render-imports';
import { METADATA_HOOK_NAME, selfClosingTags } from '../parsers/jsx';
import { MitosisComponent } from '../types/mitosis-component';
import { MitosisNode } from '../types/mitosis-node';
import { blockToReact, componentToReact } from './react';

export interface ToMitosisOptions {
  prettier?: boolean;
  format: 'react' | 'legacy';
}

export const DEFAULT_FORMAT: ToMitosisOptions['format'] = 'legacy';

// Special isValidAttributeName for Mitosis so we can allow for $ in names
const isValidAttributeName = (str: string) => {
  return Boolean(str && /^[$a-z0-9\-_:]+$/i.test(str));
};

export const blockToMitosis = (
  json: MitosisNode,
  toMitosisOptions: Partial<ToMitosisOptions> = {},
): string => {
  const options: ToMitosisOptions = {
    format: DEFAULT_FORMAT,
    ...toMitosisOptions,
  };
  if (options.format === 'react') {
    return blockToReact(json, {
      format: 'lite',
      stateType: 'useState',
      stylesType: 'emotion',
      prettier: options.prettier,
    });
  }

  if (json.name === 'For') {
    const needsWrapper = json.children.length !== 1;
    return `<For each={${json.bindings.each}}>
    {(${json.properties._forName}, index) =>
      ${needsWrapper ? '<>' : ''}
        ${json.children.map((child) => blockToMitosis(child, options))}}
      ${needsWrapper ? '</>' : ''}
    </For>`;
  }

  if (json.properties._text) {
    return json.properties._text as string;
  }

  if (json.bindings._text) {
    return `{${json.bindings._text}}`;
  }

  let str = '';

  str += `<${json.name} `;

  if (json.bindings._spread) {
    str += ` {...(${json.bindings._spread})} `;
  }

  for (const key in json.properties) {
    const value = (json.properties[key] || '')
      .replace(/"/g, '&quot;')
      .replace(/\n/g, '\\n');

    if (!isValidAttributeName(key)) {
      console.warn('Skipping invalid attribute name:', key);
    } else {
      str += ` ${key}="${value}" `;
    }
  }
  for (const key in json.bindings) {
    const value = json.bindings[key] as string;
    if (key === '_spread') {
      continue;
    }

    if (key.startsWith('on')) {
      str += ` ${key}={event => ${value.replace(/\s*;$/, '')}} `;
    } else {
      if (!isValidAttributeName(key)) {
        console.warn('Skipping invalid attribute name:', key);
      } else {
        str += ` ${key}={${value}} `;
      }
    }
  }
  if (selfClosingTags.has(json.name)) {
    return str + ' />';
  }

  // Self close by default if no children
  if (!json.children.length) {
    str += ' />';
    return str;
  }
  str += '>';
  if (json.children) {
    str += json.children
      .map((item) => blockToMitosis(item, options))
      .join('\n');
  }

  str += `</${json.name}>`;

  return str;
};

const getRefsString = (json: MitosisComponent, refs = getRefs(json)) => {
  let str = '';

  for (const ref of Array.from(refs)) {
    str += `\nconst ${ref} = useRef();`;
  }

  return str;
};

const mitosisCoreComponents = ['Show', 'For'];

export const componentToMitosis = (
  toMitosisOptions: Partial<ToMitosisOptions> = {},
): Transpiler => ({ component }) => {
  const options: ToMitosisOptions = {
    format: DEFAULT_FORMAT,
    ...toMitosisOptions,
  };

  if (options.format === 'react') {
    return componentToReact({
      format: 'lite',
      stateType: 'useState',
      stylesType: 'emotion',
      prettier: options.prettier,
    })({ component });
  }

  const json = fastClone(component);

  const refs = getRefs(json);

  mapRefs(json, (refName) => `${refName}.current`);

  const addWrapper = json.children.length !== 1;

  const components = Array.from(getComponents(json));

  const mitosisComponents = components.filter((item) =>
    mitosisCoreComponents.includes(item),
  );
  const otherComponents = components.filter(
    (item) => !mitosisCoreComponents.includes(item),
  );

  const hasState = Boolean(Object.keys(component.state).length);

  const needsMitosisCoreImport = Boolean(
    hasState || refs.size || mitosisComponents.length,
  );

  // TODO: smart only pull in imports as needed
  let str = dedent`
    ${
      !needsMitosisCoreImport
        ? ''
        : `import { ${!hasState ? '' : 'useState, '} ${
            !refs.size ? '' : 'useRef, '
          } ${mitosisComponents.join(', ')} } from '@builder.io/mitosis';`
    }
    ${
      !otherComponents.length
        ? ''
        : `import { ${otherComponents.join(',')} } from '@components';`
    }
    ${renderPreComponent(json)}

    ${
      !component.meta.metadataHook
        ? ''
        : `${METADATA_HOOK_NAME}(${json5.stringify(
            component.meta.metadataHook,
          )})`
    }

    export default function ${component.name}(props) {
      ${
        !hasState
          ? ''
          : `const state = useState(${getStateObjectStringFromComponent(
              json,
            )});`
      }
      ${getRefsString(json)}

      ${!json.hooks.onMount ? '' : `onMount(() => { ${json.hooks.onMount} })`}

      ${
        !json.hooks.onUnMount
          ? ''
          : `onUnMount(() => { ${json.hooks.onUnMount} })`
      }

      return (${addWrapper ? '<>' : ''}
        ${json.children.map((item) => blockToMitosis(item, options)).join('\n')}
        ${addWrapper ? '</>' : ''})
    }

  `;

  if (options.prettier !== false) {
    try {
      str = format(str, {
        parser: 'typescript',
        plugins: [
          require('prettier/parser-typescript'), // To support running in browsers
        ],
      });
    } catch (err) {
      console.error(
        'Format error for file:',
        str,
        JSON.stringify(json, null, 2),
      );
      throw err;
    }
  }
  return str;
};
