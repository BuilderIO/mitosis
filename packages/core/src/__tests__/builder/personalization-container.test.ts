import { componentToBuilder } from '@/generators/builder';
import { componentToMitosis } from '@/generators/mitosis';
import { builderContentToMitosisComponent } from '@/parsers/builder';
import { parseJsx } from '@/parsers/jsx';
import { describe, test } from 'vitest';
import { dedent } from '@/helpers/dedent';

describe('Builder Personalization Container/Variants', () => {
  test('Snapshot PersonalizedContainer', () => {
    const code = dedent`
      import { PersonalizationContainer, Variant } from "@components";
  
      export default function MyComponent(props) {
        return (
          <PersonalizationContainer>
            <Variant
              name="variant1"
              startDate="2024-01-01"
              query={{
                property: "urlPath",
                operation: "is",
                value: "/home",
              }}
            >
              <div>Home</div>
              <div>Div</div>
            </Variant>
            <PersonalizationOption
              name="2"
              query={[
                {
                  property: "favoriteColor",
                  operation: "is",
                  value: ["red", "blue"],
                },
              ]}
            >
              <>Red</>
            </PersonalizationOption>
            <Variant>
              <div>Default</div>
            </Variant>
            <div>More tree</div>
  
          </PersonalizationContainer>
        );
      }
    `;
  
    const component = parseJsx(code);
    const builderJson = componentToBuilder()({ component });
    expect(builderJson.data?.blocks?.[0]).toMatchSnapshot();
  
    const backToMitosis = builderContentToMitosisComponent(builderJson);
    const mitosis = componentToMitosis({
        format: 'legacy',
      })({
      component: backToMitosis,
    });
    expect(mitosis.trim()).toMatchInlineSnapshot(`
      "import { PersonalizationContainer, Variant, Fragment } from \\"@components\\";

      export default function MyComponent(props) {
        return (
          <PersonalizationContainer>
            <Variant
              name=\\"variant1\\"
              startDate=\\"2024-01-01\\"
              query={[
                {
                  property: \\"urlPath\\",
                  operation: \\"is\\",
                  value: \\"/home\\",
                },
              ]}
            >
              <div>Home</div>
              <div>Div</div>
            </Variant>
            <Variant
              name=\\"2\\"
              query={[
                {
                  property: \\"favoriteColor\\",
                  operation: \\"is\\",
                  value: [\\"red\\", \\"blue\\"],
                },
              ]}
            >
              <Fragment>Red</Fragment>
            </Variant>
            <Variant default=\\"\\">
              <div>Default</div>
              <div>More tree</div>
            </Variant>
          </PersonalizationContainer>
        );
      }"
    `);
  });
  
  test('Regenerate PersonalizedContainer', () => {
    const code = dedent`
      import { PersonalizationContainer, Variant } from "@components";
  
      export default function MyComponent(props) {
        return (
          <PersonalizationContainer>
            <Variant
              name="2"
              startDate="2024-01-01"
              endDate="2024-01-31"
              query={[
                {
                  property: "favoriteColor",
                  operation: "is",
                  value: "red",
                },
              ]}
            >
              <div>Red</div>
            </Variant>
            <Variant default="">
              <div>Default</div>
            </Variant>
          </PersonalizationContainer>
        );
      }
    `;
  
    const component = parseJsx(code);
    const builderJson = componentToBuilder()({ component });
    const backToMitosis = builderContentToMitosisComponent(builderJson);
    const mitosis = componentToMitosis({
        format: 'legacy',
      })({
      component: backToMitosis,
    });
    expect(mitosis.trim()).toEqual(code.trim());
  });
  
  test('do not generate empty variant expression', () => {
    const builderJson = {
      "@type": "@builder.io/sdk:Element" as const,
      "@version": 2,
      "id": "builder-a12265b5892b4d6e8e37873369218409",
      "component": {
        "name": "PersonalizationContainer",
        "options": {
          "variants": []
        }
      }
    }
    const backToMitosis = builderContentToMitosisComponent({ data: { blocks: [builderJson] } });    
    const mitosis = componentToMitosis()({
      component: backToMitosis,
    });
    expect(mitosis).toMatchInlineSnapshot(`
      "import { PersonalizationContainer, Variant } from \\"@components\\";

      export default function MyComponent(props) {
        return (
          <PersonalizationContainer variants={}>
            <Variant default=\\"\\" />
          </PersonalizationContainer>
        );
      }
      "
    `);
  });
});
