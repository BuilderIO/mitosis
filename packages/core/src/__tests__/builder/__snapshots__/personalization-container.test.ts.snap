// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Builder Personalization Container/Variants > Snapshot PersonalizedContainer 1`] = `
{
  "@type": "@builder.io/sdk:Element",
  "actions": {},
  "children": [
    {
      "@type": "@builder.io/sdk:Element",
      "actions": {},
      "bindings": {},
      "children": [
        {
          "@type": "@builder.io/sdk:Element",
          "bindings": {},
          "component": {
            "name": "Text",
            "options": {
              "text": "Default",
            },
          },
          "tagName": "span",
        },
      ],
      "code": {
        "actions": {},
        "bindings": {},
      },
      "properties": {},
      "tagName": "div",
    },
    {
      "@type": "@builder.io/sdk:Element",
      "actions": {},
      "bindings": {},
      "children": [
        {
          "@type": "@builder.io/sdk:Element",
          "bindings": {},
          "component": {
            "name": "Text",
            "options": {
              "text": "More tree",
            },
          },
          "tagName": "span",
        },
      ],
      "code": {
        "actions": {},
        "bindings": {},
      },
      "properties": {},
      "tagName": "div",
    },
  ],
  "code": {
    "actions": {},
    "bindings": {},
  },
  "component": {
    "name": "PersonalizationContainer",
    "options": {
      "variants": [
        {
          "blocks": [
            {
              "@type": "@builder.io/sdk:Element",
              "actions": {},
              "bindings": {},
              "children": [
                {
                  "@type": "@builder.io/sdk:Element",
                  "bindings": {},
                  "component": {
                    "name": "Text",
                    "options": {
                      "text": "Home",
                    },
                  },
                  "tagName": "span",
                },
              ],
              "code": {
                "actions": {},
                "bindings": {},
              },
              "properties": {},
              "tagName": "div",
            },
            {
              "@type": "@builder.io/sdk:Element",
              "actions": {},
              "bindings": {},
              "children": [
                {
                  "@type": "@builder.io/sdk:Element",
                  "bindings": {},
                  "component": {
                    "name": "Text",
                    "options": {
                      "text": "Div",
                    },
                  },
                  "tagName": "span",
                },
              ],
              "code": {
                "actions": {},
                "bindings": {},
              },
              "properties": {},
              "tagName": "div",
            },
          ],
          "name": "variant1",
          "query": [
            {
              "@type": "@builder.io/core:Query",
              "operation": "is",
              "property": "urlPath",
              "value": "/home",
            },
          ],
          "startDate": "2024-01-01",
        },
        {
          "blocks": [
            {
              "@type": "@builder.io/sdk:Element",
              "actions": {},
              "bindings": {},
              "children": [
                {
                  "@type": "@builder.io/sdk:Element",
                  "bindings": {},
                  "component": {
                    "name": "Text",
                    "options": {
                      "text": "Red",
                    },
                  },
                  "tagName": "span",
                },
              ],
              "code": {
                "actions": {},
                "bindings": {},
              },
              "component": {
                "name": "Core:Fragment",
              },
            },
          ],
          "name": "2",
          "query": [
            {
              "@type": "@builder.io/core:Query",
              "operation": "is",
              "property": "favoriteColor",
              "value": [
                "red",
                "blue",
              ],
            },
          ],
        },
      ],
    },
  },
}
`;
